File: src/python/mofapipeline/io.py
from __future__ import annotations

import logging
from pathlib import Path

import pandas as pd

from .config import PipelineConfig, resolve_path
from .models import MultiOmicsDataset, OmicsLayer


LOGGER = logging.getLogger(__name__)


def read_omics_layer(
    path: str | Path,
    view_name: str,
    id_column: str = "sample",
    label_token: str = "label",
) -> OmicsLayer:
    csv_path = Path(path)
    raw_df = pd.read_csv(csv_path)

    if id_column not in raw_df.columns:
        raise ValueError(f"Column '{id_column}' not found in {csv_path}.")

    label_rows = raw_df.loc[raw_df[id_column] == label_token]
    if label_rows.shape[0] != 1:
        raise ValueError(
            f"Expected exactly one label row in {csv_path}, found {label_rows.shape[0]}."
        )

    feature_df = raw_df.loc[raw_df[id_column] != label_token].copy()
    sample_ids = [col for col in feature_df.columns if col != id_column]

    if feature_df[id_column].duplicated().any():
        raise ValueError(f"Duplicated feature identifiers detected in {csv_path}.")
    if len(sample_ids) != len(set(sample_ids)):
        raise ValueError(f"Duplicated sample identifiers detected in {csv_path}.")

    for sample_id in sample_ids:
        feature_df[sample_id] = pd.to_numeric(feature_df[sample_id], errors="coerce")

    matrix = feature_df.set_index(id_column)
    metadata = pd.DataFrame(
        {
            "sample_id": sample_ids,
            "group": label_rows.iloc[0][sample_ids].astype(str).to_list(),
            "view": view_name,
        }
    )

    LOGGER.info(
        "Loaded %s from %s with %s features and %s samples.",
        view_name,
        csv_path,
        matrix.shape[0],
        matrix.shape[1],
    )

    return OmicsLayer(
        name=view_name,
        matrix=matrix,
        sample_metadata=metadata,
        source_path=csv_path.resolve(),
    )


def align_multiomics_layers(layers: list[OmicsLayer]) -> MultiOmicsDataset:
    sample_sets = [set(layer.matrix.columns) for layer in layers]
    common_samples = sorted(set.intersection(*sample_sets))
    if not common_samples:
        raise ValueError("No shared samples were found across omics layers.")

    aligned_views: dict[str, pd.DataFrame] = {}
    aligned_meta: list[pd.DataFrame] = []
    for layer in layers:
        aligned_views[layer.name] = layer.matrix.loc[:, common_samples].copy()
        meta = (
            layer.sample_metadata.loc[layer.sample_metadata["sample_id"].isin(common_samples)]
            .set_index("sample_id")
            .loc[common_samples]
            .reset_index()
        )
        aligned_meta.append(meta)

    reference = aligned_meta[0][["sample_id", "group"]].reset_index(drop=True)
    for current in aligned_meta[1:]:
        current_cmp = current[["sample_id", "group"]].reset_index(drop=True)
        if not reference.equals(current_cmp):
            raise ValueError("Sample group labels are not consistent across omics layers.")

    return MultiOmicsDataset(views=aligned_views, sample_metadata=reference)


def load_multiomics_data(config: PipelineConfig) -> MultiOmicsDataset:
    layer_specs = {
        "proteomics": config.paths["proteomics"],
        "lipidomics": config.paths["lipidomics"],
        "metabolomics": config.paths["metabolomics"],
    }
    layers = [
        read_omics_layer(
            resolve_path(config, relative_path),
            view_name=view_name,
            id_column=config.data["id_column"],
            label_token=config.data["label_token"],
        )
        for view_name, relative_path in layer_specs.items()
    ]
    dataset = align_multiomics_layers(layers)
    dataset.config_path = config.path
    return dataset

File: src/r/io.R
read_omics_layer <- function(path, view_name, id_column = "sample", label_token = "label") {
  raw_tbl <- readr::read_csv(path, show_col_types = FALSE)

  if (!id_column %in% names(raw_tbl)) {
    stop(sprintf("Column '%s' not found in %s.", id_column, path))
  }

  label_row <- raw_tbl %>%
    dplyr::filter(.data[[id_column]] == label_token)

  if (nrow(label_row) != 1) {
    stop(sprintf("Expected exactly one label row in %s, found %s.", path, nrow(label_row)))
  }

  feature_tbl <- raw_tbl %>%
    dplyr::filter(.data[[id_column]] != label_token) %>%
    dplyr::mutate(dplyr::across(-dplyr::all_of(id_column), readr::parse_number))

  raw_sample_ids <- setdiff(names(feature_tbl), id_column)
  sample_ids <- paste0("sample_", raw_sample_ids)
  feature_ids <- feature_tbl[[id_column]]

  if (anyDuplicated(feature_ids)) {
    stop(sprintf("Duplicated feature identifiers detected in %s.", path))
  }

  mat <- feature_tbl %>%
    tibble::column_to_rownames(id_column) %>%
    as.matrix()
  storage.mode(mat) <- "double"
  colnames(mat) <- sample_ids

  sample_metadata <- tibble::tibble(
    sample_id = sample_ids,
    group = as.character(unlist(label_row[1, raw_sample_ids, drop = TRUE]))
  )

  list(
    view = view_name,
    matrix = mat,
    sample_metadata = sample_metadata,
    raw_dimensions = dim(raw_tbl)
  )
}

merge_sample_metadata <- function(layer_list) {
  combined_metadata <- purrr::map_dfr(layer_list, function(layer) {
    layer$sample_metadata %>%
      dplyr::mutate(source_view = layer$view)
  })

  conflicts <- combined_metadata %>%
    dplyr::distinct(sample_id, group) %>%
    dplyr::count(sample_id, name = "n_groups") %>%
    dplyr::filter(n_groups > 1)

  if (nrow(conflicts) > 0) {
    stop("Conflicting sample labels were found across omics layers.")
  }

  combined_metadata %>%
    dplyr::group_by(sample_id) %>%
    dplyr::summarise(
      group = dplyr::first(group),
      available_views = paste(sort(unique(source_view)), collapse = ";"),
      .groups = "drop"
    ) %>%
    dplyr::arrange(sample_id)
}

collect_multiomics_layers <- function(layer_list) {
  list(
    views = purrr::set_names(purrr::map(layer_list, "matrix"), purrr::map_chr(layer_list, "view")),
    sample_metadata = merge_sample_metadata(layer_list)
  )
}

load_multiomics_data <- function(config) {
  paths <- config$paths
  settings <- config$data

  layers <- list(
    read_omics_layer(paths$proteomics, "proteomics", settings$id_column, settings$label_token),
    read_omics_layer(paths$lipidomics, "lipidomics", settings$id_column, settings$label_token),
    read_omics_layer(paths$metabolomics, "metabolomics", settings$id_column, settings$label_token)
  )

  multiomics <- collect_multiomics_layers(layers)
  multiomics$config <- config
  multiomics
}
